HDDS-10239. Storage Container Reconciliation. (#6121)
apache/ozone: hadoop-hdds/docs/content/design/container-reconciliation.md

   - From the user perspective, containers labelled quasi-closed and unhealthy represent compromised durability and availability, regardless of the container's actual contents.

2. **Focus on Recovery Paths**: Focusing on the path to a failed state is secondary to focusing on the path out of failed states.
    - For example, we should not focus on whether it is possible for two replicated closed containers with locally matching chunk checksums to have differing content, only on whether the system could detect and recover from this case if it were to happen.

3. **System Safety**:  If a decision made by software will make data more durable a single trigger is sufficient. If a decision can potentially reduce durability of data or execute an unsafe operation (unlink, trim, delete) then the confidence level has to be high, the clarity of the decision precise and clear and preferably the decision is made within services that have a wider view of the cluster (SCM/Recon).

4. **Datanode Simplicity**: Datanodes should only be responsible for safe decisions and eager to make safe choices, avoiding unsafe autonomy.

## Assumptions

> - A closed container will not accept new blocks from clients.

Reviewer: That is the long term direction, states will only hint about certain SCM actions when SCM needs to quickly pick a replica to recover from a disk failure. Another train of thought is to avoid closing containers entirely but that is out of scope for this doc.

> - A closed container will not accept new blocks from clients.

Reviewer 2: I agree that it is a long term thing and not in the scope of this.  However, we should design something today preventing the long term direction.  Let me ask a different question -- why we need this assumption in this design?

> - A closed container will not accept new blocks from clients.

Reviewer: That is right, as long as containers reconcile it should be ok for them receive new blocks. We don't need this assumption to be true in the design.

- Add a mechanism for datanodes to reconcile their replica of a container with another datanode's replica so that both replicas can be verified to be equal at the end of the process.

- Add a mechanism for SCM to trigger this reconciliation as part of the existing heartbeat command protocol SCM uses to communicate with datanodes.

- Add an `ozone admin container reconcile <container-id>` CLI that can be used to manually resolve diverged container states among non-open container replicas.

### Phase II (out of scope for this document)

- Automate container reconciliation requests as part of SCM's replication manager.

- Simplify SCM replication manager decommission and recovery logic based on mismatch of container checksums, instead of the combination of all possible container states.

### Phase III (out of scope for this document)

- Extend container level checksum verification to erasure coded containers.
    - EC container replicas do not have the same data and are not expected to have matching container checksums.
    - EC containers already use offline recovery as a reconciliation mechanism.

## Solution Implementation

### Storage

> The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):

Reviewer: We should not limit the number of blocks in a container to avoid container explosion due to small files. We can run some rough numbers to see the tradeoff between block count vs hash tree size.

> The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):

Reviewer 2: For tiny blocks, say less than 1k, we may write the block data to rocksdb (in the future) instead of storing it as a file.  This can avoid creating a lot of small files on disk.  We should think about how to optimize the checksum tree.  This may not be a big problem since container size is only 5GB.  The checksum tree size should be < 1GB.

> The only extra information we will need to store is the container merkle tree on each datanode container replica. The current proposal is store this separately as a proto file on disk so that it can be copied over the network exactly as stored. The structure would look something like this (not finalized, for illustrative purposes only):

Reviewer 3: If any block is deleted later,  will it require recalculate the tree and update the on disk file?

```
ContainerChecksum: {
  algorithm: CRC32
  checksum: 12345
  repeated BlockChecksum
}
BlockChecksum: {
  algorithm: CRC32
  checksum: 12345
  length: 5

>   deleted: false

Reviewer: We currently assume that there is no block deletion by mistake as the path for deletion of block comes from OM receiving block deletion. Thus, a block is marked as deleted only when OM has deleted the object and the blocks referenced. Protection against accidental deletes should be done via snapshots.

>   deleted: false

Reviewer 2: The might not be a valid assumption since, when there is a corruption, it is invalid to assume "something will never be corrupted".  A possible cause of block deletion by mistake is a (future) software bug.

We could have another checksum for the deleted blocks.  Below is a quick idea:
1. the deleted-block-checksum `sum` initially is 0;
2. When a block `b` is deleted, update `sum` with xor, i.e. `sum ^= hash(b)`.

SCM could store a deleted-block-checksum for each container.  Then, it is easy to tell if the deleted blocks are correct in a particular container.

>   deleted: false

Reviewer: A couple of issues
1. The flag `deleted` in code should be true only if SCM issued the deletion.
2. SCM tracking a lineage of hash changes has other memory bloat challenges for SCM which will provide fiction for scaling.
3. If SCM has a software bug that erroneously sends deletes or Datanode erroneously marks blocks as deleted, protection via MT might not be enough.
4. The only way to make sure that during reconciliation we know for sure that a block is deleted at the SCM level (not just peer) would introduce a lot of bloat to SCM.

There are invariants in this design that depend on correctness of implementation
1. Two chunks written by Ratis to two Datanodes cannot do both
    1. Agree with the checksum provided by the client
    2. Disagree with the peer Datanode on the checksum
2.  SCM decides when a block is deleted and it informs the Datanodes (followers) on the decision to delete a block. This is the only path to delete a block in a Datanode and for it to be marked as deleted on the Datanode. 
    1. If SCM sends a delete command to one Datanode (follower) it is ok for the other followers to learn about it via peer to peer communication.

>   deleted: false

Reviewer 2: > 2. SCM tracking a lineage of hash changes has other memory bloat challenges for SCM which will provide fiction for scaling.

This is just xor checkout sum over block id hash (not block data hash) for deleted blocks.  The overhead should be similar to an integer deleted block count.

> 3. ...  SCM has a software bug ...

A software bug could be in datatnodes: one datanode has mistakenly marked a block as deleted.  SCM reconciles the container with other datanodes.  During reconciliation, the other datanodes will also make that block as deleted.

>   deleted: false

Reviewer: Not sure I follow the recommendation for how to track deletions beyond adding a flag to indicate it is deleted or how the guarantees would be more robust.

>   deleted: false

Reviewer 2: We have two checksums `mtChecksum` and `deletedBlockChecksum`.  When SCM issue a reconciliation, it sends both the checksums to the datanodes:
- When the `mtChecksum` is matched, it means all the block data are correct.
  - `mtChecksum` mismatched means that some data is corrupted.
- When the `deletedBlockChecksum` is matched, it means the deleted block list is good.
   - `deletedBlockChecksum ` mismatched means that the deleted block list is incorrect.

Without `deletedBlockChecksum`, `mtChecksum` matched means both all the block data correct and deleted block list correct since it won't verify the block list.

>   deleted: false

Reviewer: Why split the logic? If a Container checksum does not match
1. Check if the container have the same blocks
    1. If blocks differ fix blocks
2. Check if any blocks have been deleted on the peer.

The additional checksum to track the sum of deleted blocks does not lead to any simplification of the runtime processing and adds a fork in the logic at the root vs. down the tree on a per block basis.

>   deleted: false

Reviewer 2: > Why split the logic? If a Container checksum does not match

It is for the other case.  If we don't split it, the deleted list could be different when container checksums are matched.

>   deleted: false

Reviewer 2: It could hit the problem described in https://github.com/apache/ozone/pull/6121#discussion_r1488638589

> A software bug could be in datatnodes: one datanode has mistakenly marked a block as deleted. SCM reconciles the container with other datanodes. During reconciliation, the other datanodes will also make that block as deleted.

>   healthy: true

Reviewer: Here is what the actual proto change will look like 
```
```diff
 diff --git a/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto b/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 index 718e2a108c..ed254478e9 100644
 --- a/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 +++ b/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 @@ -382,6 +382,7 @@ message ChunkInfo {
    repeated KeyValue metadata = 4;
    required ChecksumData checksumData =5;
    optional bytes stripeChecksum = 6;
 +  optional bool healthy = 7;
  }
  
  message ChunkInfoList {
 @@ -525,3 +526,28 @@ service IntraDatanodeProtocolService {
    rpc download (CopyContainerRequestProto) returns (stream CopyContainerResponseProto);
    rpc upload (stream SendContainerRequest) returns (SendContainerResponse);
  }
 +
 +message MTBlockInfo { // Merkle Tree Block Info
 +  optional BlockData blockData = 1; // The chunks in this should be sorted by order of chunks written.
 +  optional ChecksumData checksumData = 2; // Checksum of the checksum of the chunks.
 +  optional bool deleted = 3; // True if the block is deleted.
 +  optional int64 length = 4; // Length of the block
 +  optional int64 chunkCount = 5; // Number of chunks in the block.
 +}
 +
 +message MTContainerInfo {
 +  // Merkle Tree Container Info
 +  enum FailureCause {
 +    NO_HEALTHY_CHECKSUM_FOUND_WITH_PEERS = 1; // No healthy checksum found with peers.
 +    NO_PEERS_FOUND = 2; // No peers found.
 +  }
 +  optional ContainerDataProto containerData = 1;
 +  repeated MTBlockInfo mtBlockInfo = 2; // Sorted list of blocks.
 +  optional ChecksumData checksumData = 3; // Checksum of the checksum of the blocks
 +  optional bool deleted = 4; // True if the container is deleted.
 +  optional int64 blockCount = 5; // Number of blocks in the container.
 +  optional FailureCause failureCause = 6; // Failure cause if the container is unhealthy. Only used for reporting.
 +  optional int64 reconciliationCount = 7; // Number of times reconciliation was attempted.
 +}
 ```

>   healthy: true

Reviewer: Please take a look at the updated proto definition.

  repeated ChunkChecksum
}
ChunkChecksum: {

>   algorithm: CRC32

Reviewer: We plan to reuse the existing checksum objects.
```
```diff
 diff --git a/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto b/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 index 718e2a108c..ed254478e9 100644
 --- a/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 +++ b/hadoop-hdds/interface-client/src/main/proto/DatanodeClientProtocol.proto
 @@ -382,6 +382,7 @@ message ChunkInfo {
    repeated KeyValue metadata = 4;
    required ChecksumData checksumData =5;
    optional bytes stripeChecksum = 6;
 +  optional bool healthy = 7;
  }
  
  message ChunkInfoList {
 @@ -525,3 +526,28 @@ service IntraDatanodeProtocolService {
    rpc download (CopyContainerRequestProto) returns (stream CopyContainerResponseProto);
    rpc upload (stream SendContainerRequest) returns (SendContainerResponse);
  }
 +
 +message MTBlockInfo { // Merkle Tree Block Info
 +  optional BlockData blockData = 1; // The chunks in this should be sorted by order of chunks written.
 +  optional ChecksumData checksumData = 2; // Checksum of the checksum of the chunks.
 +  optional bool deleted = 3; // True if the block is deleted.
 +  optional int64 length = 4; // Length of the block
 +  optional int64 chunkCount = 5; // Number of chunks in the block.
 +}
 +
 +message MTContainerInfo {
 +  // Merkle Tree Container Info
 +  enum FailureCause {
 +    NO_HEALTHY_CHECKSUM_FOUND_WITH_PEERS = 1; // No healthy checksum found with peers.
 +    NO_PEERS_FOUND = 2; // No peers found.
 +  }
 +  optional ContainerDataProto containerData = 1;
 +  repeated MTBlockInfo mtBlockInfo = 2; // Sorted list of blocks.
 +  optional ChecksumData checksumData = 3; // Checksum of the checksum of the blocks
 +  optional bool deleted = 4; // True if the container is deleted.
 +  optional int64 blockCount = 5; // Number of blocks in the container.
 +  optional FailureCause failureCause = 6; // Failure cause if the container is unhealthy. Only used for reporting.
 +  optional int64 reconciliationCount = 7; // Number of times reconciliation was attempted.
 +}
 ```

  checksum: 12345
  length: 5
  offset: 5
  healthy: true
}
```

- The `deleted` flag in the block proto allows it to serve as a tombstone entry so that block deletion does not affect the container level hash.

- The `algorithm` field is the algorithm used to aggregate hashes from the layer below which results in that object's current hash.
    - We can define an order of chunks and blocks based on their IDs so that the hash does not need to be commutative (order independent).

- The `healthy` field in the chunk checksum being true indicates that the computed chunk checksum used to fill in the merkle tree matches the chunk checksum stored in RocksDB.

> - The `healthy` field in the block checksum being true indicates that all chunks within the block are marked as `healthy = true`.

Reviewer: Yes, the container checksum would reflect the health of the block. As this would mean a chunk was not included in calculating the block checksum.

> - We may be able to re-use the `ChunkInfo` and `ChecksumData` messages in `DatanodeClientProtocol.proto` for chunk information and checksums. Other structures in the merkle tree storage will require new objects to be defined.

Reviewer: I considered that, the problem is that we cannot reduce size without introducing a whole bunch of new APIs or paying for extra roundtrips. Today we have the infrastructure to create chunk readers and start reading chunks. If we do not pass in the existing structures in the MT, we would either need new APIs to read chunk or have additional round trips. We will experiment with this and see if we can optionally not set some of the fields or cheaply introduce new on the wire structures. One option is to complete https://issues.apache.org/jira/browse/HDDS-10338 as part of this work which would make the proto definitions a lot leaner.

## APIs

The following APIs would be added to datanodes to support container reconciliation. The actions performed when calling them is defined in [Events](#events).

- `reconcileContainer(containerID, List<Replica>)`
    - Instructs a datanode to reconcile its copy of the specified container with the provided list of replica's containers.
    - Datanodes will call `getContainerHashes` for each replica to identify repairs needed, and use existing chunk read/write APIs to do repairs necessary.
    - This would likely be a new command as part of the SCM heartbeat protocol, not actually a new API.
- `getContainerHashes(containerID)`
    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Storage](#storage)

## Events

This section defines how container reconciliation will function as events occur in the system.

### Reconiliation Events

 These events occur as part of container reconciliation in the happy path case.

#### On Data Write

> - No change to existing datanode operations. Existing chunk checksums passed and validated from the client will be persisted to RocksDB.

Reviewer: We could do it as part of last put block but that is an optimization we can visit once the basic framework is in place. Doing it offline allows writes to not pay the cost of calculating the hash tree.

#### On Container Close

> - Container checksum is calculated synchonously from existing checksums. This calculation must finish before the close is acked back to SCM.

Reviewer: We can, it would be best to do this once we can measure the performance hit of doing it as part of write or for an open container.

> - Container checksum is calculated synchonously from existing checksums. This calculation must finish before the close is acked back to SCM.

Author: This will be tricky to implement, because multiple blocks can be written to a container in parallel and the checksum will require the same order across all nodes. Block IDs are handed out to clients in an increasing order, but there is no guarantee that the client(s) write them to the container in that order.

    - The Merkle Tree is still updated to reflect the current contents of the disk in this case.

#### On Container import

- Schedule on-demand scan after the container import completes.
    - There may be cases where we need to import a slightly corrupt container since it is our best option. Therefore, do not do a synchronous scan of container and fail the import if corruption is detected.
    - Instead, schedule the imported container for on-demand scanning to be sure SCM knows of corruption if it does not already.
    - **Optimization**: Clear last scanned timestamp on container import to make the background scanner get to it faster if datanode restarts before the on-demand scan completes.

#### On `getContainerHashes`

- Return the current Merkle Tree of the specified container.
    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.
- If container not closed, return error.
- If container not found, return error.

> - The merkle tree returned by this call will always be consistent.

Reviewer 2: The proto file could be corrupted.  We need to checksum it. :)

Reviewer: Good point. we can also store a hash within the tree that represents the has of the content excluding the hash itself.

Reviewer: Will cover this in the implementation for the Merkle Tree

    - The hash of each node in the tree will always accurately represent a hash of its children.
- The merkle tree returned by this call may not always match the current state of the container.
    - The container can be repaired or corrupted any time between the scan that generated the tree and the call to read the container's tree.
    - If the caller decides it needs to read a chunk based on the merkle tree obtained, it will have to re-check the checksum obtained as part of the chunk read to verify the data received matches what it was expecting.

#### On `reconcileContainer`

- Use existing merkle tree, unless its not there, then compute it.
    - The tree will be calculated from existing chunk checksums if it is not present. However, this should not happen in the majority of cases since it will be calculated on container close.

> 1. For each replica:

Reviewer 4: What about finding missed deletes? Is that something we want to address too?

In the case of a block missing from one replica (or an additional block in one replica). MTs are ordered structures, so if you are missing a block or have an extra block in one MT "early" in the sorted order, it will effectively make all the hashes differ between the two trees and we will need to fall back to a block by block comparison. How do we handle that, or is my understanding incorrect?

Reviewer: Yes there would be a linear walk of the blocks to reconcile which is not different from a width first walk of a tree.

    1. Datanode calls `getContainerHashes`
    2. Datanode diffs this with its current merkle tree to find all chunks that are:
        1. Not part of blocks that are marked as deleted in this or the other replica's tree.
        2. Not matching the chunk's checksum value in this datanode's tree.
    3. For each chunk in the other replica's tree that does not match this replica's tree:
        1. If the chunk or block is not present in our container, save it in memory as needing to be imported.
        2. If the chunk or block is not marked `healthy` in our merkle tree (known to be corrupted by the scanner), save it in memory as needing to be reconciled.
        3. If chunk checksums differ but our chunk is marked `healthy`, ignore this chunk.
            - If the chunk is corrupted in the other replica or got corrupted while we were reading it, we should not import it.
            - If the chunk is corrupted on one of the replicas but the `healthy` flag has not been updated yet to reflect this, it will be resolved when the scanner reaches the container again.

>             - If the same chunk is truly `healthy` in both replicas but the checksums do not match, there has been an irrecoverable error on the write path where different data was written to each replica but they both passed checksum validation.

Reviewer 4: We probably need a way to flag such a block and exclude it from the container hash. Otherwise the replica hashes will always different and the container will always been seen as "needs reconciliation" from SCMs perspective.

Reviewer: Good point.

Reviewer: SCM co-ordination is not in scope for this document for now.

#### On Client read of container under reconciliation

- No action required.

This read could be an end client or another datanode asking for data to repair its replica. The result will only be inconsistent if the chunk/block is being repaired. In this case the data is already corrupt and the client will get a read checksum failure either way. Even if all datanodes are reconciling the same container at the same time, this should not cause an availability problem because there has to be one correct replica that others are repairing from that the client can also read.

#### Replicate command for a container being reconciled

- Allow replication to proceed with best effort, possibly with a warning log.

There may be cases where a container is critically under-replicated and we need a good copy of the data as soon as possible. Meanwhile reconciliation could remained blocked for an unrelated reason. We can provide a way to pause/unpause reconciliation between chunk writes to do reasonably consistent container exports.

> #### On block delete

Reviewer 4: Its not clear to me how this mechanism can detect "lost deletes" where a block should be deleted from all replicas, but someone the block got left behind on one or more of them.

If the MT ignores the deleted blocks so that when they are deleted the container hash does not change, then is there any efficient way to detect lost deletes?

Lost block deletes are something we want to address, and it is something we should design into this solution too.

Say, on delete, we update a flag in the MT, but do not adjust the container hash. We cannot detect lost deletes at SCM based on the root hash alone. Therefore nothing would trigger reconciliation unless there is an additional corruption in a replica to trigger it.

I agree that the container hash should not be changed due to a delete, but I think we many need something else to highlight lost delete problems - eg a deleted hash (combination of checksums of deleted block) and possibly a deleted block list or deleted block MT. I am further unsure if a MT could help find a lost delete, as they are not suited for proof of membership / non membership AFAIK. Sorted block lists may be need to be compared.

Reviewer: The Merkle Tree records blocks that have been deleted. The hash of the container does not change but the Merkle Tree's metadata for the block changes.

5. Deleted blocks within a container that still exists in some container replicas.
6. The need to simplify the replication manager for how to handle cases where only quasi-closed and unhealthy container replicas are available.

Ideally, a healthy Ozone cluster would contain only open and closed container replicas. However, container replicas commonly end up with a mix of states including quasi-closed and unhealthy that the current system is not able to resolve to cleanly closed replicas. The cause of these states is often bugs or broad failure handling on the write path. While we should fix these causes, they raise the problem that Ozone is not able to reconcile these mismatched container replica states on its own, regardless of their cause. This has lead to significant complexity in the replication manager for how to handle cases where only quasi-closed and unhealthy replicas are available, especially in the case of decommissioning.

Even when all container replicas are closed, the system assumes that these closed container replicas are equal with no way to verify this. During writes a client provides a checksum for the chunk that is written. 

> The scanner validates periodically that the checksums of the chunks on disk match the checksums provided by the client. It is possible that the checksum of a chunk on disk does not match the client provided checksum recorded at the time of write. Additionally, during container replica copying the consistency of the data is not validated, opening the possibility of silent data corruption propagating through the system.

Reviewer: ```suggestion
The scanner validates periodically that the checksums of the chunks on disk match the checksums provided by the client. It is possible that the checksum of a chunk on disk does not match the client provided checksum recorded at the time of write. Additionally, during container replica copying, the consistency of the data is not validated, opening the possibility of silent data corruption propagating through the system.
```

This document proposes a container reconciliation protocol to solve these problems. After implementing the proposal:

> 1. It should be possible for a cluster to progress to a state where it has only properly replicated closed and open containers.

Reviewer: ```suggestion
1. It should be possible for a cluster to progress to a state where all not open containers are closed and meeting the desired replication factor.
```

+}

```
This is written as a file to avoid bloating the RocksDB instance which is in the IO path.

## APIs

The following APIs would be added to datanodes to support container reconciliation. The actions performed when calling them is defined in [Events](#events).

- `reconcileContainer(containerID, List<Replica>)`
    - Instructs a datanode to reconcile its copy of the specified container with the provided list of other container replicas.
    - Datanodes will call `getContainerHashes` for each container replica to identify repairs needed, and use existing chunk read/write APIs to do repairs necessary.
    - This would likely be a new command as part of the SCM heartbeat protocol, not actually a new API.
- `getContainerHashes(containerID)`
    - A datanode API that returns the merkle tree for a given container. The proto structure would be similar to that outlined in [Merkle Tree](#Container-Hash-Tree-/-Merkle-Tree).

## Reconciliation Process

SCM: Storage Container Manager

DN: Datanode

### SCM sets up the reconciliation process as follows:

1. SCM triggers a reconciliation on `DN 1` for container 12 with replicas on `DN 2` and `DN 3`.

>    1. `SCM -> reconcileContainer(Container #12, DN2, DN3) -> DN 1` 

Reviewer 5: Do we need to consider the issue of repeated sending, especially after HA switch or SCM restart?

Author: Yes. We can handle this similarly to how retries or duplicate commands of container replication are handled on datanodes currently. I can add more details to the doc on this.

Reviewer: Duplicate commands can be deduplicated on the Datanode side. For SCM ordering containers it will need to have some prioritization for containers as they are reported via heartbeats.

>    1. `SCM -> reconcileContainer(Container #12, DN2, DN3) -> DN 1` 

Author: In Phase I of development, which this document covers, the  reconcile request is triggered manually by a CLI command given to SCM. In Phase II of development, reconciliation will be triggered when the SCM replication manager sees that the container checksums of multiple containers do not match.

>    1. `SCM -> reconcileContainer(Container #12, DN2, DN3) -> DN 1` 

Reviewer: In addition to what @errose28 said, SCM can choose to schedule a reconciliation even for containers that have no checksum reported. Example: One UNHEALTHY and a second that is QUASI_CLOSED. At first DNs would only calculate the Merkle Tree and report back and then SCM can decide further. As pointed out by @errose28 we will cover the SCM automation in a separate design doc.

### Datanodes set up the reconciliation process as follows:

> 1. `DN 1` schedules the Merkle Tree to be calculated if it is not already present and report it to SCM. 

Author: Yes, we should unify these. Since we plan on calculating the merkle tree synchronously on close, the case of a container not having the tree should only be for containers created before an upgrade that the scanner has not reached yet.

   1. SCM can then compare the container replica hashes and schedule reconciliation if they are different.
2. If `DN 1` already has the Merkle Tree locally, it will compare it with the Merkle Trees of the other container replicas and schedule reconciliation if they are different. Example: 
   1. `DN 1 -> getContainerHashes(Container #12) -> DN 2` // Datanode 1 gets the merkle tree of container 12 from Datanode 2.
   2. `DN 1 -> getContainerHashes(Container #12) -> DN 3` // Datanode 1 gets the merkle tree of container 12 from Datanode 3.
   3. ... // Continue for all replicas.

### Reconcile loop once the merkle trees are obtained from all/most replicas:

> 1. `DN 1` checks if any blocks are missing. For each missing Block:

Reviewer 3: What's the criteria to determine whether a block is missing, or merely an extra not needed block?  The question is what's the rule to decide which replica is the baseline replica.

Reviewer: 1. The reconciliation process does not take into account if a block is referenced or not by OM. If the block was written to a container it will be replicated across as a safety measure. Recon can at a later point identify unwanted blocks.
2. Yes when adding a block the metadata for it will also need to be updated.

4. `DN 1` recomputes Merkle tree and sends it to SCM via ICR (Incremental Container Report).

Note: This document does not cover the case where the checksums recorded at the time of write match the chunks locally within a Datanode but differ across replicas. We assume that the replication code path is correct and that the checksums are correct. If this is not the case, the system is already in a failed state and the reconciliation protocol will not be able to recover it. Chunks once written are not updated, thus this scenario is not expected to occur.

#### `getContainerHashes`

When a datanode receives a request to get container hashes. The following steps are performed:

1. If the Merkle Tree is present, return it.
2. If the Merkle Tree is not present, the call `getContainerHashes` will return and error. 
   1. Datanode missing the Merkle Tree will be schedule the scanner to calculate the Merkle Tree and then report to SCM.
3. If container not closed, return error.
4. If container not found, return error.
5. The merkle tree returned by this call will represent the status of the data on disk (post scanner scrub).

> ## Sample scenarios

Reviewer 6: This might have discussed somewhere above, but I got the following question:
1. What do we do when only one container available in quasi closed state?
2. When only 2 quasi closed containers are available.
3. When 2 quasi closed containers are available and post reconciliation, if old closed container comes back. ( we discussed this in offline chat, but it would be good to capture those scenarios)
With # 2 like cases, how do we avoid falsely treating container as good? ( if we reconcile from 2 quasi-closed states, then it may not be guaranteed we have all chunks right?) Could we please discuss those scenarios and clarify. Thanks

Reviewer: We will dive deeper into the state changes for container post reconciliation in the SCM doc around reconciliation.
You raise a valid point that is beneficial to remember if a container was `QUASI_CLOSED` if it cannot be compared to a `CLOSED` container.

1. **Container is missing blocks**
   1. `DN 1` has 10 blocks, `DN 2` has 11 blocks, `DN 3` has 12 blocks.
   2. `DN 1` will read the missing block from `DN 2` and `DN 3` and store it locally.
   3. `DN 1` will recompute the merkle tree and send it to SCM.

2. **Container has corrupted chunks**

>    1. `DN 1` has block 20: chunk 13, `DN 2` has block 12: chunk 13

Reviewer: Yes. This is to highlight that both DNs have corrupted blocks but when reconciliation is on going on `DN1` it will only fix its local chunk that is corrupt. `DN2` when it gets a reconcile command will reconcile its corrupted chunk.

Since the scanner can generate the container merkle trees in the background, existing containers created before this feature will still be eligible for reconciliation. These old containers may not have all of their block deletes present in the merkle tree, however, which could cause some false positives about missing blocks on upgrade if one node had already deleted blocks from a container before the upgrade, and another datanode has not yet processed the delete of those blocks.

This can be mitigated by:
1. Having datanodes delete blocks from their container replica on reconciliation that another replica has marked as deleted.
2. Having SCM periodically ask datanodes to reconcile otherwise matching containers when there are no other mismatched containers in the cluster.

The sequence of events would look like this. Assume software v1 does not have container reconciliation, but v2 does.
1. Datanode 1 in v1 deletes block 1 in container 1. This does not leave a tombstone entry because v1 does not have container reconciliation.
2. Cluster is upgraded to v2.
3. Reconciliation is triggered between datanodes 1 and 2 for container 1. Datanode 2 has not yet deleted block 1.

> 4. Datanode 1 will add block 1 to its container when reconciling with datanode 2.

Reviewer 5: If the `delTX` of Datanode 2 is greater than Datanode 1, Datanode 1 can confirm that there is no need to restore Block 1 by deleting the tombstone of Datanode 2.
So maybe you can do this by waiting for all replicas' `delTX` to be greater than a certain value (perhaps the largest `delTX` among all replicas at the beginning of reconciliation)

    - If we stop here, then block 1 will remain orphaned on datanode 1.
5. Datanode 2 deletes block 1 and adds a tombstone entry for it since the cluster is running software v2 which has container reconciliation.
6. Since container checksums disregard deleted blocks, container 1 will be seen as matching from SCM's point of view. However, periodic reconciliation requests for closed containers will still eventually ask these two replicas to reconcile.

> 7. Datanode 1 learns that block 1 was deleted from datanode 2, so it moves the block metadata to the deleted table in RocksDB

Author: There is a `bool deleted` flag in the merkle tree proto that will be updated for each block delete as a tombstone entry. When datanodes use the proposed API to pull the Merkle tree, they can see that their peer deleted the block so it is not missing.

> 7. Datanode 1 learns that block 1 was deleted from datanode 2, so it moves the block metadata to the deleted table in RocksDB

Reviewer 5: Understood. Datanode 1 needs to wait for SCM to send the "reconcileContainer" API call to it. Once this call is received, Datanode 1 will retrieve the Merkle tree from Datanode 2. This allows Datanode 1 to know that block 1 has been deleted, right?
However, this process may be time-consuming, as SCM might not sending "reconcileContainer" solely due to an orphan block.

> 7. Datanode 1 learns that block 1 was deleted from datanode 2, so it moves the block metadata to the deleted table in RocksDB

Reviewer: @xichen01 this is good point, I think similar to a node level scan we will need a SCM level scan where every container in the system is reconciled once in `x` number of days even if the checksums match. The other option is what @szetszwo recommended is to have a separate delete hash in the Merkle Tree to detect missing deletions.

> 7. Datanode 1 learns that block 1 was deleted from datanode 2, so it moves the block metadata to the deleted table in RocksDB

Author: Yes we should not depend on reconcile purely for deleting blocks. Main block delete flow will still be active. There are corner cases where the current delete flow can leave orphan blocks, so it would be nice to have for the reconciliation process to handle this, although it is not a primary goal.

>     - A delete transaction entry with an ID key would need to be created to do this. Currently these are only received from SCM and not created by the datanode.

Author: Right now the delete IDs are created from SCM's sequence ID generator. There was previously an assumption that the delete transaction IDs would be processed by the datanode in ascending order but I think we removed that check.

Pros of re-using the delete table instead of deleting the block directly is that the block deleting service code path is reused. Cons of not deleting the block directly when reconciling is that the unique delete transaction ID can now come from two different sources.
